staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1264,31 +1274,29 @@ func (c *Cacher) getBookmarkAfterResourceVersionLockedFunc(ctx context.Context,
 //   - start at Any (return the current watchCache's RV)
 //
 // The returned function must be called under the watchCache lock.
-func (c *Cacher) getStartResourceVersionForWatchLockedFunc(ctx context.Context, parsedWatchResourceVersion uint64, opts storage.ListOptions) (func() uint64, error) {
+func (c *Cacher) getStartResourceVersionForWatchLockedFunc(parsedResourceVersion, watchCacheResourceVersion uint64, opts storage.ListOptions) (func() uint64, error) {
 	if opts.SendInitialEvents == nil || *opts.SendInitialEvents {
-		return func() uint64 { return parsedWatchResourceVersion }, nil
+		return func() uint64 { return parsedResourceVersion }, nil
 	}
-	return c.getCommonResourceVersionLockedFunc(ctx, parsedWatchResourceVersion, opts)
-}
-
-// getCommonResourceVersionLockedFunc a helper that simply computes a ResourceVersion
-// based on the input parameters. Please examine callers of this method to get more context.
-//
-// The returned function must be called under the watchCache lock.
-func (c *Cacher) getCommonResourceVersionLockedFunc(ctx context.Context, parsedWatchResourceVersion uint64, opts storage.ListOptions) (func() uint64, error) {
-	switch {
-	case len(opts.ResourceVersion) == 0:
-		rv, err := storage.GetCurrentResourceVersionFromStorage(ctx, c.storage, c.newListFunc, c.resourcePrefix, c.objectType.String())
-		if err != nil {
-			return nil, err
-		}
-		return func() uint64 { return rv }, nil
-	case parsedWatchResourceVersion == 0:
+	if parsedResourceVersion == 0 {
 		// here we assume that watchCache locked is already held
 		return func() uint64 { return c.watchCache.resourceVersion }, nil
-	default:
-		return func() uint64 { return parsedWatchResourceVersion }, nil
 	}
+	return func() uint64 { return watchCacheResourceVersion }, nil
+}
+
+// getWatchCacheResourceVersion returns a ResourceVersion to which the watch cache must be synchronized to
+//
+// Depending on the input parameters, the semantics of the returned ResourceVersion are:
+//   - must be at Exact RV (when parsedWatchResourceVersion > 0)
+//   - can be at Any RV (when parsedWatchResourceVersion = 0)
+//   - must be at Most Recent RV (return an RV from etcd)
+func (c *Cacher) getWatchCacheResourceVersion(ctx context.Context, parsedWatchResourceVersion uint64, opts storage.ListOptions) (uint64, error) {
+	if opts.SendInitialEvents == nil || !*opts.SendInitialEvents || !opts.Predicate.AllowWatchBookmarks || len(opts.ResourceVersion) != 0 {
```

Comment:
I'm not sure this condition is actually correct here...

For the purpose of this condition the fact if we should send initial events or not - doesn't matter.
The only thing that matters here is what RV and ResourceVersionMatch were requested by the users.

RV="" NotOlderThan  => RV from storage
RV="" Exact => we can't guarantee that?? this should fail validation??
RV="0" NotOlderThan => return 0
RV="0" Exact => this should fail validation
RV="value" NotOlderThan => return parsedRV
RV="value" Exact => this should fail validation

Reply:
I think it is correct because the validation says that the `match` must be `NotOlderThan` and that the `match` is required when you set `SendInitialEvents` (https://github.com/kubernetes/kubernetes/blob/2ae115e5a2c9b0ce3be8b3f4b60f8acc1b7c9108/staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go#L53)

I think that this code (as it is now) is consistent with the other checks we implemented in `cacher.go`, e.g. we check the `SendInitialEvents` whenever we need strong consistency (a call to etcd).

If we skip checking `SendInitialEvents` then I think we might potentially change the semantic for `RV=""` (no `match` parameter). But because there is an additional check in the `waitUntilWatchCacheFreshAndForceAllEvents` method we won't change the semantic if we wished to remove checking the `SendInitialEvents`. So do you want to remove checking `SendInitialEvents` ?

Reply:
OK - so at the very least let's start with a very explicit comment what validation allows:
- if SendInitiaEvents != nil => ResourceVersionMatch = NotOlderThan
- if ResourceVersionmatch != nil => ResourceVersionMatch = NotOlderThan & SendInitialEvents != nil

Now, the question is, what should happen if SendInitialEvents == nil - we should do what we were doing before, being not waiting for anything.

But if AllowWatchBookmarks is false - is doesn't matter, we need to maintain the requested semantic - in this case we simply just don't send bookmark but everything else is the same.

Reply:
[Before](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L599) this PR the considered `requestedWatchRV`.  This semantics stays the same for this PR when `SendInitialEvents == nil` or `AllowWatchBookmarks is false`.  I can rename `parsedWatchResourceVersion` to `requestedWatchRV` if that helps.

Reply:
Which seems like a bug to me.

If SendInitialEvents == true and ResourceVersion = ""

the I don't care if the bookmark should be send or not - we need to start from the RV being current etcd from RV.

Reply:
OK, I think you are right `SendInitialEvents == true and ResourceVersion = ""` is our legacy case (https://github.com/kubernetes/kubernetes/blob/52d613efde52437a64a1d4438f28e50d59c1596e/staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/defaults.go#L32)